pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -430,6 +438,164 @@ func consumerRefMatches(consumer *corev1.ObjectReference, machine *machinev1.Mac
 	return true
 }
 
+// updateUserData fetches full ignition from MCS and updates the UserData secret reference
+func (a *Actuator) updateUserData(ctx context.Context, config *bmv1alpha1.BareMetalMachineProviderSpec,
+	machine *machinev1.Machine) error {
+
+	// Try to fetch the full ignition secret to see if it exists
+	machineUserDataSecret := &corev1.Secret{}
+	key := client.ObjectKey{
+		Name:      machine.Name + "-user-data",
+		Namespace: machine.Namespace,
+	}
+	// Check if the full ignition secret already exists. If not, create one
+	err := a.client.Get(ctx, key, machineUserDataSecret)
+	if errors.IsNotFound(err) {
```

Comment:
This test really begins the "happy path" through this code, but with that logic nested inside this conditional it isn't immediately obvious. The function would be clearer if real errors were handled first with a quick return. That would change this to something like

```
if err == nil {
    return nil
}
if !errors.IsNotFound(err) {
    log.Printf("Cannot get %s secret: %w", key.Name, err)
    return err
}
```

and then everything between lines 454 and 562 would be shifted left one level, outside of a conditional statement.

Doing that also makes clear that we are assuming that an existing secret contains the correct content. Do we want to check that and update it if it is out of date?

Reply:
This is *exactly* what @mhrivnak pointed out earlier. I'm in the process of making this change. I guess great minds think alike.

Reply:
At the moment, we're not checking whether the ignition out of date because the original data secret reference was assigned to the host while setting its spec. The userdata secret was assumed (I think, not sure) to be static and I don't know where, if at all, it is modified at a later time. All this PR's function is doing is changing the secret reference to a new secret made from full ignition.